### PR TITLE
Refresh tokens without search

### PR DESCRIPTION
**Pull request by albertzaharovits**
This introduces a new access and refresh token format 
in order to avoid the token doc search when using refresh tokens
and instead rely on realtime-get.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:762]: @slobodanadamovic As far as I can tell, at least till v8.10, this change should by itself be enough to update the mapping of the tokens index automatically (the code is hard to follow, and I'm not sure what the plans post 8.10 are, so I would appreciate if you could double check it, please).

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:762]: Here's the work tracking how it should work post 8.10: https://github.com/elastic/elasticsearch/pull/97529

Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecuritySystemIndices.java:762]: Thanks for checking it and linking the PR. 👍  I've only had brief information that there was a plan to migrate system indices away from version. Hence why I mentioned it. But seems like we are still depending on it and the mapping update will be handled automatically.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2076]: nit: We know how much bytes we will write. I suggest we pass in the expected size to the `BytesStreamOutput` constructor in order to avoid relocation of the underlying byte array.

**Review by slobodanadamovic — COMMENTED** [review]
I have yet to review tests. Here are some comments for the main code. Nothing blocking so far.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: I know the source parsing is a madness, but it's only done in this one place. I'm not sure doing some parsing in a constructor for `AccessAndRefreshTokens` would help.
However, I've pushed [03f6084](https://github.com/elastic/elasticsearch/pull/97395/commits/03f60845dbb3b9497bc5999ef448a465fca5ca6c) to address the parsing madness. It removes some of the nesting, to avoid going over the line length limit which results in dubious formatting.
Let me know if this is looking more reasonable now.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: You mean API docs like this one https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-get-token.html ? It's only the length that's differing with the new tokens, from the user's perspective. Maybe I'll handle it after we merge this one.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: Pushed [2a2e7f7](https://github.com/elastic/elasticsearch/pull/97395/commits/2a2e7f75e6a1fc6d3ce09fca80ffc89f614937d0) to define `RAW_TOKEN_BYTES_TOTAL_LENGTH = RAW_TOKEN_BYTES_LENGTH + RAW_TOKEN_DOC_ID_BYTES_LENGTH`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: Also handled in https://github.com/elastic/elasticsearch/pull/97395/commits/2a2e7f75e6a1fc6d3ce09fca80ffc89f614937d0 .

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: Looks much better. 👍

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: Yes, that was my initial thought. But on the second thought, you are right. There is nothing special about the tokens except the length. Feel free to ignore above comment.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:1102]: I stared at this wondering if we should still keep tests which are actually testing with UUIDs for BWC?
Reason I'm not sure if it's beneficial here is because we already have `TokenBackwardsCompatibilityIT` tests. WDYT?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:1056]: good nit: pushed [d76a62d](https://github.com/elastic/elasticsearch/pull/97395/commits/d76a62dfc74187837169fa677d7a6ef72c87fdea)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2129]: Yeah, I agree the usage of both `Base64.getUrlEncoder()` and `Base64.getEncoder()` is confusing and a possible source of bugs...
So, I think I can make base64 encoding uniform (use the non-URL-safe one), at least for post 7.2 tokens (I plan to remove pre 7.2 support altogether in a follow-up) though this slightly exceeds the scope of this PR, I will help a lot with readability...
Let me know if you're against fixing it in this PR.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:1102]: I think it's still OK and useful to be able to have a `TokenService` instance that can be made to understand only specific token formats and also be able to feed it different ones. We can easily test the format parsing this way. The `TokenBackwardsCompatibilityIT` is very hard to troubleshoot (I had to do it when commiting the mistake from [44a9ad2](https://github.com/elastic/elasticsearch/pull/97395/commits/44a9ad27c93f2e102490fee9122f886f3ce9869d)), although it is indeed more comprehensive because it does not rely on mocking, so it tests everything including that token documents are stored and retrieved in the expected form.

I plan to remove pre-7.2 tokens in a follow-up. It would've been easier to do that before this change, but it would've taken more time to land the new format feature, and I preferred to prioritize the work that worried me more (adding new functionality rather than removing existing one).

Hope the above answers your point?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2224]: Indeed I favored simplicity. I'll raise the one that does the data copying. We'll discuss over these.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2129]: Thanks for explanation. I'm fine either way. Do what's easier for you.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:1102]: Makes sense. Thanks.
I only have n additional suggestion to add `test-full-bwc` label to this PR. I think it is useful given the changes.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:1102]: Good tip! I'm going to also mark the PR as "ready-for-review" as I'm not sure full-bwc tests trigger for draft PRs.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java]: I've pushed [c7f774f](https://github.com/elastic/elasticsearch/pull/97395/commits/c7f774f8e42ae7a794ab541e97bf2e0bd78c74a0) . It looks like this option is equally simple.

Looks like we're left with a choice between:
 * throwing away 8 securely generated bytes, OR
 * allocating an extra array of 16 bytes and writing and copying from it

What's your preference? I don't have a good grasp of the performance difference between the two.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2225]: Oops, pushed fix [3a7186b](https://github.com/elastic/elasticsearch/pull/97395/commits/3a7186b0572657d46dca352805a36e047f46d11b)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2129]: I was wrong. It's not possible to converge the base64 encoding schemes.... I can only change the encoding used in the new token version. The old tokens, as well as the encrypted superseding tokens, must continue to use whatever encoding/decoding scheme was already in place.
Is it worth changing regular base64 encoding over the url-safe one for new tokens only? I'm oscillating...

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2225]: > What's your preference? I don't have a good grasp of the performance difference between the two.

I've made a simple local test to compare the two implementations. 
It turns out that the second option (_allocating an extra array of 16 bytes and writing and copying from it_) is slightly faster. On average it's faster around 35ms than generating extra 8 bytes which are thrown away. I've tested it running `100 000` generations. With `10 000` generations, the difference is ~2-3ms.
Even though the difference is almost negligible, I'd prefer the second option.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:113]: After looking closer, I don't see what's the purpose of the `testHashedTokenIsUrlSafe` test method.
It tests the `TokenService#hashTokenString` method, whose output is only stored in token docs, not as part of URLs...
I'm going to remove this test.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2225]: WOW, thanks for testing!!!

Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2225]: Just to make it clear, `100 000` generations with option 1 is slower than option 2 by ~35ms. So for every `100 000` generations we save ~ 35ms. But what I did not consider is that this costs ~1.5 MB of RAM which has to be garbage collected. I have no idea how much time GC would take and it may be that in the end we break even.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2225]: Right, it's tricky to benchmark durations that also involve the GC as the GC is influenced by the other activity in the VM.

**Comment by albertzaharovits** [comment]
@slobodanadamovic I've only added some unit tests for malformed tokens: [7ec895b](https://github.com/elastic/elasticsearch/pull/97395/commits/7ec895b5a0ff4f76d6bb3cc74b9470822a9685f1) . These are important because they cover the path where the _token doc id_ part of a received token (access and refresh) is correct, but the part associated to the stored hash is wrong, which is a different behavior introduced by the new token format.

**Comment by albertzaharovits** [comment]
Otherwise, I think the existing (unit) tests cover it enough.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2129]: 🤔 Looking at the RFCs ([6750](https://datatracker.ietf.org/doc/html/rfc6750#section-2.1) and updated [9110](https://datatracker.ietf.org/doc/html/rfc9110#name-authentication-parameters)) they don't state that the token has to be `base64` nor `base64Url` encoded. It just states that it has to support characters which allow usage of `base64` and `base64Url` encoded tokens. So I think we are fine in both cases. I'll defer to you for the decision.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:112]: Right. I initially read it as if this is what's being returned. I'm okay if we remove it.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TokenBackwardsCompatibilityIT.java
Comment (slobodanadamovic) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TokenBackwardsCompatibilityIT.java:346]: Good catch! 👍

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:2129]: Tokens returned to users are always `base64` encoded (see eg https://github.com/elastic/elasticsearch/blob/b504346ff4bb76b81087ea39fbcd9c59af4a2889/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L1938). The issue here is that stuff that gets stored in token docs sometimes uses `base64` and sometimes `base64Url`. And since we can't get to a uniform convention because of bwc concerns, I'm leaning to leave it as is.

**Review by slobodanadamovic — APPROVED** [review]
LGTM 🚀 

Thanks for explanations and iterations.

**Comment by albertzaharovits** [comment]
Thanks for the quick iterations!

**Pull request merged**
